File: backend/app/schemas/stock.py
"""Pydantic schemas for stock data endpoints."""

from pydantic import BaseModel

from app.schemas.common import MetaResponse


class StockResponse(BaseModel):
    """Flat representation of a Stock record."""

    symbol: str
    name: str
    market: str
    industry: str | None = None
    description: str | None = None

    model_config = {"from_attributes": True}


class PricePointResponse(BaseModel):
    """Single OHLCV bar."""

    date: str
    open: float
    high: float
    low: float
    close: float
    volume: int
    adj_close: float | None = None


class IndicatorPointResponse(BaseModel):
    """Single date/value pair for a technical indicator series."""

    date: str
    value: float


class ValuationResponse(BaseModel):
    """Fundamental valuation metrics for a stock.

    All fields default to ``None`` when the data source does not provide them
    (e.g. Taiwan stocks for US-only fields).
    """

    # Core valuation
    pe_ratio: float | None = None
    pb_ratio: float | None = None
    ps_ratio: float | None = None
    dividend_yield: float | None = None
    market_cap: float | None = None
    eps: float | None = None
    revenue: float | None = None
    profit_margin: float | None = None

    # 風險指標
    beta: float | None = None
    fifty_two_week_high: float | None = None
    fifty_two_week_low: float | None = None

    # 財務健康
    debt_to_equity: float | None = None
    current_ratio: float | None = None
    quick_ratio: float | None = None

    # 獲利能力
    roe: float | None = None
    roa: float | None = None
    operating_margin: float | None = None
    gross_margin: float | None = None
    free_cash_flow: float | None = None

    # 成長指標
    revenue_growth: float | None = None
    earnings_growth: float | None = None

    # 進階估值
    peg_ratio: float | None = None
    ev_to_ebitda: float | None = None
    forward_pe: float | None = None

    # 分析師
    target_mean_price: float | None = None
    recommendation_key: str | None = None
    number_of_analysts: float | None = None

    # 持股結構
    insider_holding: float | None = None
    institutional_holding: float | None = None

    # 做空資料
    short_ratio: float | None = None
    short_percent_of_float: float | None = None

    # 股息詳情
    payout_ratio: float | None = None
    dividend_rate: float | None = None
    five_year_avg_dividend_yield: float | None = None


class StockListResponse(BaseModel):
    """Paginated list of stocks."""

    data: list[StockResponse]
    meta: MetaResponse


class StockDetailResponse(BaseModel):
    """Single stock detail envelope."""

    data: StockResponse | None


class StockHistoryResponse(BaseModel):
    """Historical OHLCV data envelope."""

    data: list[PricePointResponse]


class StockIndicatorsResponse(BaseModel):
    """Technical indicators envelope.

    ``data`` is a dict keyed by indicator name.  Simple single-series
    indicators (sma, ema, rsi) map to ``list[IndicatorPointResponse]``.
    Multi-series indicators (macd, bollinger_bands, kd) map to a nested
    dict where each sub-key is also a ``list[IndicatorPointResponse]``.
    """

    data: dict[str, list[IndicatorPointResponse] | dict[str, list[IndicatorPointResponse]]]


class StockValuationResponse(BaseModel):
    """Valuation metrics envelope."""

    data: ValuationResponse


class StockSignal(BaseModel):
    """A single qualitative signal from the scoring engine."""

    type: str  # "positive" | "negative" | "neutral"
    message: str


class StockScoreResponse(BaseModel):
    """Composite quality score for a stock."""

    overall_score: int
    valuation_score: float
    technical_score: float
    fundamental_score: float
    grade: str
    signals: list[StockSignal]


class StockScoreDetailResponse(BaseModel):
    """Stock score envelope."""

    data: StockScoreResponse

File: frontend/src/types/stock.ts
export interface Stock {
  symbol: string;
  name: string;
  market: 'US' | 'TW';
  industry?: string;
  description?: string;
  price?: number;
  change?: number;
  changePercent?: number;
  volume?: number;
}

export interface PricePoint {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  adjClose?: number;
}

export interface IndicatorPoint {
  date: string;
  value: number;
}

export interface Watchlist {
  id: number;
  name: string;
  items: { id: number; symbol: string }[];
}

export interface BacktestConfig {
  symbol: string;
  strategy: string;
  startDate: string;
  endDate: string;
  initialCapital: number;
  commission?: number;
  tax?: number;
  params?: Record<string, number>;
}

export interface BacktestResult {
  totalReturn: number;
  annualizedReturn: number;
  maxDrawdown: number;
  sharpeRatio: number;
  winRate: number;
  totalTrades: number;
  finalValue: number;
  equityCurve: { date: string; value: number }[];
  trades: { date: string; action: string; price: number; shares: number }[];
}

export interface ScreenerCondition {
  indicator: string;
  operator: 'gt' | 'gte' | 'lt' | 'lte' | 'eq';
  value: number | string;
}

export interface ValuationData {
  peRatio: number | null;
  pbRatio: number | null;
  psRatio: number | null;
  dividendYield: number | null;
  marketCap: number | null;
  eps: number | null;
  revenue: number | null;
  profitMargin: number | null;

  // 風險指標
  beta: number | null;
  fiftyTwoWeekHigh: number | null;
  fiftyTwoWeekLow: number | null;

  // 財務健康
  debtToEquity: number | null;
  currentRatio: number | null;
  quickRatio: number | null;

  // 獲利能力
  roe: number | null;
  roa: number | null;
  operatingMargin: number | null;
  grossMargin: number | null;
  freeCashFlow: number | null;

  // 成長
  revenueGrowth: number | null;
  earningsGrowth: number | null;

  // 進階估值
  pegRatio: number | null;
  evToEbitda: number | null;
  forwardPe: number | null;

  // 分析師
  targetMeanPrice: number | null;
  recommendationKey: string | null;
  numberOfAnalysts: number | null;

  // 持股結構
  insiderHolding: number | null;
  institutionalHolding: number | null;

  // 做空
  shortRatio: number | null;
  shortPercentOfFloat: number | null;

  // 股息詳情
  payoutRatio: number | null;
  dividendRate: number | null;
  fiveYearAvgDividendYield: number | null;
}

export interface StockScore {
  overallScore: number;
  valuationScore: number;
  technicalScore: number;
  fundamentalScore: number;
  grade: string;
  signals: { type: 'positive' | 'negative' | 'neutral'; message: string }[];
}

export interface ApiResponse<T> {
  data: T;
  meta?: { total: number; page: number; limit: number };
}

export type Market = 'US' | 'TW' | 'ALL';

export type AlertType =
  | 'price_above'
  | 'price_below'
  | 'rsi_above'
  | 'rsi_below'
  | 'sma_cross'
  | 'volume_above';

export interface Alert {
  id: number;
  symbol: string;
  alert_type: AlertType;
  condition: Record<string, number>;
  is_active: boolean;
  triggered_at: string | null;
  created_at: string;
}

export interface AlertCheckResult {
  triggered: boolean;
  current_value: number;
}
